Attains parameter to wqx characteristic alias table update and create the cst internal ref file

wokenny13 opened the pull request:
- Changed the name of the reference file to ATTAINSParamToWQXCharRef.
- 
- Updated the ATTAINSParamToWQXCharRef file from the WQX Characteristic Alias Table
- 
- Created the CST internal ref file. This is in a clean format to be used for future TADA functions in the future.
- 
- Created an optional review table for potential additional ATTAINS to WQX alias that could be added as additional rows to the WQX Characteristic Alias Table (these will need to be reviewed). See TADA_AdditionalCharAliasForReview(). This does not create an internal ref file of any kind. We can consider moving this function to utilities? This function uses an exact/like match logic to help identify if there are potential additional matches between ATTAINS Parameter Names and WQX Characteristic Names.

wokenny13 committed:
Updates to WQX - ATTAINS Ref file

Changed the name of the reference file to ATTAINSParamToWQXCharRef.

Updated the ATTAINSParamToWQXCharRef file from the WQX Characteristic Alias Table

Created the CST internal ref file. This is in a clean format to be used in the future.

Created an optional review for potential additional ATTAINS to WQX alias that could be added as additional rows to the WQX Characteristic Alias Table (these will need to be reviewed).

wokenny13 committed:
Update TADA_RandomTestingData.Rd

wokenny13 committed:
Merge branch 'develop' into ATTAINS-Parameter-To-WQX-Characteristic-Alias-Table-Update-and-Create-the-CST-Internal-Ref-File

wokenny13 committed:
Update TADA_AdditionalCharAliasForReview Documentation

see examples for output dataframe which shows potential additional alias matches.

added @export

added draft test-that to ensure ATTAINS param domain and CST pollutant name domain are up to date in the ref file

wokenny13 committed:
minor roxygen2 documentation fix in TADA_AdditionalCharAliasForReview

@example to @examples

hillarymarler commented on R/ATTAINSCrosswalks.R:
At line 754 "By default this
#' function will generate a user friendly Excel spreadsheet that includes a
#' drop down list list of all ATTAINS parameters that are applicable to the
#' organization selected by the function input 'org_id'."

This is unclear. What does "that are applicable to the organization" mean in this context? Is this previously listed parameter causes only? Or is there some other mechanism to determine which parameters apply? It also looks like there may be a typo with "list" repeated twice.

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,362 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
-  # If there is a cached table available return it
-  if (!is.null(ATTAINSParameterWQPCharRef_Cached)) {
-    return(ATTAINSParameterWQPCharRef_Cached)
+TADA_GetATTAINSParamToWQPCharRef <- function(charAliasType = c("All", "ATTAINS")) {
```
Add details about this param in the documentation?

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,362 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
-    return(ATTAINSParameterWQPCharRef_Cached)
+TADA_GetATTAINSParamToWQPCharRef <- function(charAliasType = c("All", "ATTAINS")) {
+  
+  charAliasType <- match.arg(charAliasType)
```
I can run the whole function successfully and return the crosswalk, but if I run through step by step, I get this error here: 

<img width="565" height="54" alt="image" src="https://github.com/user-attachments/assets/a4ab0fba-7005-4c2d-89e4-40034ef76cb5" />

wokenny13 replied:
this occurs throughout other TADA functions that uses the match.arg function. I am unsure if this is an expected error or not. 

Ex. in ResultFlagsDependent.R
<img width="630" height="77" alt="image" src="https://github.com/user-attachments/assets/470496ee-090f-4ef8-97ba-0e4e980f9387" />

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,362 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+  ATTAINSParamRef <- ATTAINS.raw[, "name", drop = FALSE]
+  
+  # Create the initial ATTAINS param to WQX char crosswalk
+  if(charAliasType == "ATTAINS") {
```
The condition has length > 1 error appears in both the "ATTAINS" and "All" sections too.

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,362 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+#' Criteria Search Tool Reference Table
+#'
+#' Function downloads and returns the newest available criteria search tool from 
+#' OST. Table is formatted and cleaned up for easy to  
```
This sentence looks unfinished

hillarymarler committed:
suggested removal of intermediate objects in TADA_GetCriteriaSearchToolRef

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+  CriteriaSearchToolRef_Cached <- CST
+
+  # remove intermediate objects
+  rm(CST.raw, first_filled_row_index, CST.cols)
```
I added this - suggested removal of intermediate objects.

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+  ATTAINSParamToWQPCharRef <- ATTAINSWQX2.0
+
+  ATTAINSParamToWQPCharRef
+}
```
Can some of these intermediate objects be removed before returning the ref?

wokenny13 committed:
Bug fix in TADA_AdditionalCharAliasForReview and other checks

removing intermediate variables, and add to global variables in utilities.

some minor code comment cleanups

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
 
-# Update  ATTAINS Organization Identifier Reference Table
+
+# Update CriteriaSearchToolRef Reference Table internal file
```
Do these update functions need to be added to the "Update All TADA Reference Files" section of MaintenanceScheduled.R?

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+#' of ATTAINS.ParameterNames and WQX CharacteristicNames to the logic.
+#' used in this function, which looks at the percentage of words that are
+#' a match between each ATTAINS parameters and WQX Characteristics (and
+#' within ATTAINS and CST, and CST and WQX by CAS) as an alternative
```
This seems like a separate process from the word matching? It might be more clear for users to list this as a clear additional step, rather than in parentheses next to the word match description.

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+#' used in this function, which looks at the percentage of words that are
+#' a match between each ATTAINS parameters and WQX Characteristics (and
+#' within ATTAINS and CST, and CST and WQX by CAS) as an alternative
+#' way of finding additional aliases. It is recommended for the TADA
```
Should running this function be added to scheduled maintenance? Or does it not make sense to do that since manual review is required?

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+#' accurate, and if so, reach out to the WQX team to add these to the
+#' WQX Characteristic Alias table.
+#'
+#' Note for Development: We should keep a reference file to indicate
```
Is there a draft of this internal ref file?

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+#' less strict may result in greater number of matches that shouldn't
+#' be matched (false positives). Default for now is to be more strict.
+#'
+#' @param includeCST a Boolean value. If TRUE, this will include columns
```
I'm not sure what "contains an alias between any 3 sources" is referring to here. Does it mean if any match between WQP characteristic or ATTAINS parameter with a CST pollutant is found, the CST info will be included?

wokenny13 replied:
If an alias is found between

1) ATTAINS and WQX
2) ATTAINS and CST

we can include an alias table between the 3 sources.

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+#' review <- TADA_AdditionalCharAliasForReview()
+#'
+TADA_AdditionalCharAliasForReview <- function(includeCST = FALSE
+                                              # ATTAINS.CST.tolerance,
```
Are these additional params to be added later?

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+  # The current WQX char and ATTAINS Parameter alias table from the WQX
+  ATTAINSParamToWQPCharRef <- utils::read.csv(system.file("extdata", "ATTAINSParamToWQPCharRef.csv", package = "EPATADA"))
+
+  # ATTAINSWQX2.0_missing <- ATTAINSParameterWQPCharRef |>
```
Will this section be utilized later? Maybe add a comment describing what it will be used for?

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+  WQXCharacteristicRef$CAS.Number <- gsub("-","", WQXCharacteristicRef$CAS.Number)
+
+  WQX_CST_CAS_Ref <- WQXCharacteristicRef |>
+    dplyr::inner_join(CST, by = c("CAS.Number" = "CAS_NO")) |>
```
Displays warning message that an unexpected many-to-many relationship is found. Should this be set to relationship = "many-to-many"?

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+    ) |>
+    dplyr::slice_max(order_by = percent_match_WQX + percent_match_ATTAINS ) |>
+    dplyr::right_join(WQXCharacteristicRef, by = "CharacteristicName") |>
+    dplyr::filter(
```
Would it make more sense to filter for either percent_match_WQX or percent_match_ATTAINS equal to some value (less than 1 but still fairly high)?

wokenny13 replied:
I have included this as an optional param now to address for tolerance between 0 and 1 (0% and 100%)

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+      percent_match_WQX + percent_match_ATTAINS > 1
+    )
+
+  # optional additional filter, skip if we are fine with many to many matches
```
If this is optional should there be a param added to choose it?

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+    ) |>
+    dplyr::slice_max(order_by = percent_match_WQX + percent_match_ATTAINS )
+
+  # more aggressive (too strict, can lead to missed matched)
```
Should this be commented out or made an option with a param?

wokenny13 replied:
It seems like I have defaulted to using a more aggressive approach to the crosswalk matches. I am unsure if we would like to adjust or play around with different tolerance levels to see how many additional or less matches are found, but kept it at a single value for consistency for now.

Ex.
`  temp_100 <- temp |>
    dplyr::filter(
      percent_match_WQX == 1 | percent_match_ATTAINS == 1
    )
`

the **== 1** indicates that there must be a 100% match from either end (comparing an ATTAINS parameter name to a WQX char name or vice versa) for it to be matched.

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+      percent_match_WQX == 1 | percent_match_ATTAINS == 1
+    )
+
+  temp_ATTAINS_CST <- dplyr::right_join(CST2, ATTAINSParamRef2, by = "name_words") |>
```
Another unexpected "many-to-many" relationship warning.

wokenny13 committed:
removal of additonal intermediate variables not needed throughout function

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+    )
+
+  # Join by ATTAINS Parameter name and CAS numbers.
+  temp_final <- temp_100 |>
```
unexpected "many-to-many" relationship warning

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -9,91 +9,365 @@ ATTAINSParameterWQPCharRef_Cached <- NULL
+  return(ATTAINSWQX2.0_non_matched3)
+}
+
+
```
@wokenny13 - this is a great function! The different steps you use to match/compare are logical and well thought out. I think some intermediate objects can be removed from this one too during/at the end of the workflow. Also, a few more comments on what is happening in some of the intermediate tables generated might be helpful for future maintenance and review.

hillarymarler commented on tests/testthat/test-ATTAINSCSTRefTables.R:
```diff
@@ -0,0 +1,40 @@
+  
+  # Find the first row that has all values populated. This will indicate the column names of the CST data frame.
+  # Note: Why not use a static row number? The CST may get new entries that may change the start of the data frame's. 
+  first_filled_row_index <- which(rowSums(is.na(CST.raw)) == 0)[1]
```
Would it make sense to write any of this as an internal function so it can be used in both the function and this test?

hillarymarler reviewed (changes requested):
I left comments on a few minor things - warning messages for many-to-many relationships, intermediate objects, etc.

I'm excited to see all of this incorporated into the mod 3 workflow!

hillarymarler committed:
Update test-ATTAINSCSTRefTables.R

add quiet to rExpertQuery function used in test

cristinamullin committed:
Merge branch 'develop' into ATTAINS-Parameter-To-WQX-Characteristic-Alias-Table-Update-and-Create-the-CST-Internal-Ref-File

cristinamullin committed:
Merge branch 'develop' into ATTAINS-Parameter-To-WQX-Characteristic-Alias-Table-Update-and-Create-the-CST-Internal-Ref-File

wokenny13 committed:
minor text fixes

cristinamullin committed:
Merge branch 'develop' into ATTAINS-Parameter-To-WQX-Characteristic-Alias-Table-Update-and-Create-the-CST-Internal-Ref-File

wokenny13 committed:
edits to documentation

wokenny13 committed:
Changed CST extdata name to EPACST

The current placeholder for EPA304a criteria and methods using the CST. Changed the extdata name to EPACST to clarify this.

+minor text fixes

wokenny13 committed:
Update TADA_AdditionalCharAliasForReview.Rd

wokenny13 committed:
code cleanup and documentation update

condensed and cleaned up code and updated documentation (examples and params) in TADA_AdditionalCharAliasForReview

moved TADA CriteriaSearchTool to CriteriaRefTables.R

cristinamullin committed:
Merge branch 'develop' into ATTAINS-Parameter-To-WQX-Characteristic-Alias-Table-Update-and-Create-the-CST-Internal-Ref-File

hillarymarler commented on R/ATTAINSCrosswalks.R:
```diff
@@ -750,9 +750,9 @@ TADA_UpdateATTAINSAUMLCrosswalk <- function(org_id = NULL,
+#' drop down list of all ATTAINS parameters that have been listed as a cause in 
+#' prior ATTAINS cycle for the organization selected in the function input 'org_id'. 
+#' It also highlights the cells in which users should input information. The excel
 #' spreadsheet will be automatically downloaded to a user's downloads folder path.
```
Should users have the option to specify a path for the download? (With the download folder set as the default). I'm thinking about situations where users might be using this and other TADA functions as part of their own assessment package or tool.

wokenny13 replied:
At some point in the past, it was decided to keep the path to just the downloads path. It would be an extra param argument input that would be then needed to be used throughout all of the other TADA functions, adding an extra step in the process if a user chooses to proceed with using a different path.

I am thinking it could still be useful to allow users to change the path and this is something we can discuss more in future meetings!

hillarymarler commented on R/ATTAINSCrosswalks.R:
At line 782-784: "Future development efforts may allow users to pull in magnitude values
#' for an ATTAINS parameter through the Criteria Search Tool depending on a
#' users quality control and review of these metrics."

Does this mean that future efforts will allow users to pull in magnitude values (but those values will require review). Or that we will go through a quality control/review process as part of the function development?

wokenny13 replied:
Users would be the ones required to the review process if we do pull in the magnitude values. This would be dependent on doing a crosswalk of CST pollutant names with the ATTAINS.ParameterName. The WQX Characteristic alias table in the ATTAINSParamToWQPCharRef could also contain this crosswalk if we do decide to proceed with this route in the future (make sure we choose to source from "all" to show what matches have been found for WQX Characteristics and CST pollutant names and see if an ATTAINS.Parameter is also listed as an alias for the same source.

hillarymarler commented on R/ATTAINSCrosswalks.R:
At line 789 - should we have a link to a doc listing the TADA priority characteristics?

wokenny13 replied:
I have this on line 793 edited

 'TADAPriorityChar <- utils::read.csv(system.file("extdata", "TADAPriorityCharUnitRef.csv", package = "EPATADA"))'.

hillarymarler commented on R/ATTAINSCrosswalks.R:
at 817 - could users also run TADA_GetATTAINSOrgIDsRef() to see the list?

hillarymarler committed:
Update ATTAINSCrosswalks.R

hillarymarler commented on R/ATTAINSRefTables.R:
```diff
@@ -1,99 +1,388 @@
 #' so subsequent calls will be faster.
 #'
-#' @return Updated sysdata.rda with updated ATTAINSParameterWQPCharRef object
+#' @param charAliasType A string value to indicate the WQX data source to use
```
Should there be an example of a use case where charAliasType is something other than "ATTAINS"? I'm having trouble picturing how this would look.

cristinamullin replied:
Good question. @wokenny13 are any of the other alias types useful/logical for use in this function? I bolded a few to consider.

Here is the full list of alias types from WQX:

**ATTAINS.PARAMETER**
**CAS NUMBER**
**CST.POLLUTANT**
**CST.STD.POLLUTANT****
EPA ID (SUBSTANCE REGISTRY #)
ITIS TAXON SERIAL NUMBER
MOLECULAR WEIGHT
NOAA - National Center for Environm
NWIS PARM CODE
ONTOLOGY - HYDRO.GEODAB.EU
RETIRED NAME
**STANDARDIZE NAME (Normalized)**
STORET CHARACTERISTIC NAME
STORET PARM CODE
SYSTEMATIC NAME
TAXON COMMON NAME
**WQP COMPARABLE NAME**
**WQX SYNONYM REGISTRY (validation)**

cristinamullin replied:
note: WQP COMPARABLE NAME can be populated by the TADA team to support any additional matches that would otherwise be added manually.

hillarymarler committed:
run rExpertQuery function quietly

hillarymarler reviewed (approved):
Overall, the updates look good to me. I was able to run TADA_CreateParamRef successfully on random data sets from a few different states and the output made sense.

I think we could take a look as a team at some of the documentation related to the Mod 3 functions and work to make it a little more concise and clear, especially some of the text regarding review by the TADA team or future development. But that could be done as part of a future PR.

I pushed a couple of small commits - a minor grammar change or two in documentation and the addition of spsUtil::quiet. I suggest applying that when we use rExpertQuery functions as part of larger TADA functions as the printed rExpertQuery messages are probably less useful in the greater TADA context.

wokenny13 committed:
uppercase WQX Char name in alias table

wokenny13 committed:
testthat update for TADA_CreateParamRef

No longer requires a one to one match for WQX Characteristics to ATTAINS Parameter.

wokenny13 committed:
minor updates and fixes to Mod3 Vignette

changed autofill in TADA_DefineCriteriaMethodology to "Org" for ATTAINS ParameterName to WQX Char crosswalk. We should discuss how to handle this in a future TADA team meeting.

Minor updates to the EPACST.csv (R8 seems to be working on a more up to date and reviewed EPA304s criteria table, this table has the information crosswalk from the CST for the time being)

shorten TADA mod 3 example data frame names in Mod3Vignette - AtlOptions

wokenny13 committed:
Merge branch 'develop' into ATTAINS-Parameter-To-WQX-Characteristic-Alias-Table-Update-and-Create-the-CST-Internal-Ref-File
